### ISSUE TITLE
Exceptionally Slow Training Speed

### ISSUE THREAD

**Issue by clutch-61** [TencentBAC/RoT#5]
I noticed that our actual training time is significantly longer than what is reported in the paper, and I am reaching out for help to resolve this extremely slow training issue.

The dataset (GSM8k-Aug-NL) and the model (Qwen3-VL-4B-Instruct) we selected are strictly consistent with your setup. However, I did notice one minor discrepancy: the official paper states that Stage 1 uses 1 epoch, whereas the repository configuration is set to 2 epochs.

Here are our logs:
(rot) :~/liangliu/RoT/RoT$ CUDA_VISIBLE_DEVICES=5,6 bash run_train_stage1.sh
==========================================
Image-CoT Training with DeepSpeed
==========================================
 
Configuration:
  Number of GPUs: 2
  DeepSpeed config: configs/deepspeed_config.json
  Training config: configs/stage1_config_qwen3vl_4b.yaml
  Dataset: gsm8kaug
  Batch size: 16
  Number of epochs: 2
  Learning rate: 2e-5
  Master port: 2951
 
Starting DeepSpeed training...
  MASTER_PORT=2951
  MASTER_ADDR=localhost
  CUDA_VISIBLE_DEVICES=5,6 (num_gpus=2)
[2026-08-25 21:10:59,820] [WARNING] [runner.py:232:fetch_hostfile] Unable to find hostfile, will proceed with training with local resources only.
Detected VISIBLE_DEVICES=5,6: setting --include=localhost:5,6
[2026-08-25 21:10:59,820] [INFO] [runner.py:630:main] cmd = /user_home/zhanglong/.conda/envs/rot/bin/python3.10 -u -m deepspeed.launcher.launch --world_info=eyJsb2NhbGhvc3QiOiBbNSwgNl19 --master_addr=127.0.0.1 --master_port=2951 --enable_each_rank_log=None --log_level=info scripts/train.py --config configs/stage1_config_qwen3vl_4b.yaml --dataset gsm8kaug --batch_size 16 --num_epochs 2 --lr 2e-5 --deepspeed_config configs/deepspeed_config.json
[2026-08-25 21:11:05,159] [INFO] [launch.py:162:main] WORLD INFO DICT: {'localhost': [5, 6]}
[2026-08-25 21:11:05,159] [INFO] [launch.py:168:main] nnodes=1, num_local_procs=2, node_rank=0
[2026-08-25 21:11:05,159] [INFO] [launch.py:179:main] global_rank_mapping=defaultdict(<class 'list'>, {'localhost': [0, 1]})
[2026-08-25 21:11:05,160] [INFO] [launch.py:180:main] dist_world_size=2
[2026-08-25 21:11:05,160] [INFO] [launch.py:184:main] Setting CUDA_VISIBLE_DEVICES=5,6
[2026-08-25 21:11:05,160] [INFO] [launch.py:272:main] process 2401333 spawned with command: ['/user_home/zhanglong/.conda/envs/rot/bin/python3.10', '-u', 'scripts/train.py', '--local_rank=0', '--config', 'configs/stage1_config_qwen3vl_4b.yaml', '--dataset', 'gsm8kaug', '--batch_size', '16', '--num_epochs', '2', '--lr', '2e-5', '--deepspeed_config', 'configs/deepspeed_config.json']
[2026-08-25 21:11:05,161] [INFO] [launch.py:272:main] process 2401334 spawned with command: ['/user_home/zhanglong/.conda/envs/rot/bin/python3.10', '-u', 'scripts/train.py', '--local_rank=1', '--config', 'configs/stage1_config_qwen3vl_4b.yaml', '--dataset', 'gsm8kaug', '--batch_size', '16', '--num_epochs', '2', '--lr', '2e-5', '--deepspeed_config', 'configs/deepspeed_config.json']
================================================================================
Configuration
================================================================================
Dataset: gsm8kaug
Batch size: 16
Num epochs: 2
Learning rate: 2e-05
Save interval: 50 steps
Local rank: 0
DeepSpeed config: configs/deepspeed_config.json
Device: cuda
================================================================================
  Note: NCCL_TIMEOUT not set, using default 1800 seconds
✓ TensorBoard logging to: output/logs/stage1/tensorboard
  Log directory: /user_home/zhanglong/liangliu/RoT/RoT/output/logs/stage1
 
================================================================================
Initializing Model
================================================================================
Loading weights: 100%|███████████████████████████████████████████████████████████████████████████| 713/713 [00:01<00:00, 554.65it/s]
Loading weights: 100%|███████████████████████████████████████████████████████████████████████████| 713/713 [00:01<00:00, 697.95it/s]
  Model config:
    - Use projection head: True
    - Projection hidden dim: 2048
    - Freeze vision: True
    - Enable LoRA: False
    - Vision loss weight: 10.0
    - LM loss weight: 1.0
 
================================================================================
Loading Data
================================================================================
✓ Loaded 385620 samples from data/GSM8k-Aug-NL/gsm8kaug_train_processed.jsonl
  Using Qwen3-VL chat template with EOS token: <|im_end|>
 
================================================================================
Initializing DeepSpeed
================================================================================
  DeepSpeed config: configs/deepspeed_config.json
  Local rank: 0
  World size: 2
  ZeRO Stage: 2
  CPU Offload: False
  Use custom optimizer: True
  Created custom AdamW optimizer
  Train batch size: 128
  Micro batch size per GPU: 16
  Gradient accumulation steps: 4
  Learning rate: 2e-05
[2026-08-25 21:11:30,438] [WARNING] [config_utils.py:70:_process_deprecated_field] Config parameter cpu_offload is deprecated use offload_optimizer instead
[2026-08-25 21:11:30,438] [WARNING] [config_utils.py:70:_process_deprecated_field] Config parameter cpu_offload is deprecated use offload_optimizer instead
  Found and moved model.visual.rotary_pos_emb to cuda:0
  Found and moved model.language_model.rotary_emb to cuda:0
✓ DeepSpeed initialized successfully
  Model device: cuda:0
  Vision encoder device: cuda:0
  Language model device: cuda:0
 
✓ Training setup complete
  Device: cuda:0
  Local rank: 0
  World size: 2
  Use DeepSpeed: True
  Training samples: 385620
  Batch size: 16
  Total steps: 48204
 
Logging Configuration:
  Log directory: /user_home/zhanglong/liangliu/RoT/RoT/output/logs/stage1
  TensorBoard dir: /user_home/zhanglong/liangliu/RoT/RoT/output/logs/stage1/tensorboard
  JSONL log file: /user_home/zhanglong/liangliu/RoT/RoT/output/logs/stage1/training_log.jsonl (exists)
  TensorBoard files: 6 file(s) found
 
 
================================================================================
Starting Training
================================================================================
 
================================================================================
Epoch 1/2
================================================================================
Epoch 1:   0%|                 | 49/12051 [58:14<294:34:25, 88.36s/it, loss=117.7542, v_loss=0.1396, lm_loss=116.3573, lr=1.20e-05]
[Step 50] Saving checkpoint...
 
Saving Stage 1 checkpoint (projection_head only): output/checkpoints/stage1/checkpoint_step_50
  Saving projection_head.bin (Stage 1 training)...
  ✓ Saved projection_head.bin (44.02 MB, took 0.22 seconds)
  ✓ Saved <img_begin> embedding (norm=50.7500)
  ✓ Saved <img_end> embedding (norm=50.5000)
  ✓ Saved special token embeddings (0.0112 MB)
  Saving optimizer and scheduler state for resume training...
  ✓ Saved optimizer and scheduler state (DeepSpeed checkpoint)
✓ Saved Stage 1 checkpoint to output/checkpoints/stage1/checkpoint_step_50 (took 35.60 seconds)
  Components: projection_head + special tokens + optimizer/scheduler
We are training on 2x A800 GPUs. I reviewed the logs and did not notice any obvious warnings. I captured the first 50 steps above, and after testing up to 150 steps, the iteration time remains relatively stable (around 88.36 s/it).

According to the official numbers: Stage I takes 19.0 h / 1 epoch on a single H20 GPU with a batch size of 16.
This translates to approximately 2.8 s/it (385k / 16 ≈ 24k steps). There is a massive throughput discrepancy between the official benchmark and our reproduction.

Are there any recommended optimization suggestions to alleviate this slow training speed issue? Thank you very much!

**Comment by WYFDUT** [comment]
Thanks for the very detailed report, this is helpful.  

The per-step latency gap is not caused by the epoch setting, but by the lack of batched forward passes in the model implementation. Despite a configured ``batch_size=16``, target OCR encoding, per-token autoregressive vision embedding generation, and loss computation are executed sequentially per sample inside Python loops. This forces the 4B model to run effectively at batch size 1 with excessive small-kernel and Python overheads.

Maybe you can temporarily mitigate latency by reducing the rendered CoT length/vision token count, or by lowering ``batch_size`` while increasing gradient accumulation steps to save memory without hurting throughput.

We will fix the stage 1 epoch count to 1 to match the paper, and prioritize adding real batched forward support for both ``compress_cot`` and the autoregressive vision token generation, which should give the largest speedup. Will update this issue once that lands.
